src/signals-mach.c

```diff
@@ -279,7 +279,7 @@ kern_return_t catch_mach_exception_raise(
     int nthreads = jl_atomic_load_acquire(&jl_n_threads);
     for (tid = 0; tid < nthreads; tid++) {
         jl_ptls_t _ptls2 = jl_atomic_load_relaxed(&jl_all_tls_states)[tid];
-        if (pthread_mach_thread_np(_ptls2->system_id) == thread) {
+        if (pthread_mach_thread_np(jl_atomic_load_relaxed(&_ptls2->system_id)) == thread) {
```

Comment:
```suggestion
        if (_ptls2->current_task != NULL && pthread_mach_thread_np(_ptls2->system_id) == thread) {
```

I think this might be the minimal change to fix your bug without having to mess with the existing atomics

Reply:
Implemented this in https://github.com/JuliaLang/julia/pull/50871. Thanks for the suggestion!